Allow JWT auth with Box

dbramwell opened the pull request:
#### What is the purpose of this change?

Allow passing in a Box config.json file when configuring a Box remote. Also potentially paves the way for JWT auth in other apps.

#### Was the change discussed in an issue or in the forum before?

[2582](https://github.com/rclone/rclone/issues/2582)

#### Checklist

- [x] I have read the [contribution guidelines](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#submitting-a-pull-request).
- [x] I have added tests for all changes in this PR if appropriate.
- [x] I have added documentation for the changes if appropriate.
- [x] All commit messages are in [house style](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#commit-messages).
Almost...
- [x] I'm done, this Pull Request is ready for review :-)

ncw commented on backend/box/box.go:
```diff
@@ -84,6 +107,13 @@ func init() {
 			Name: config.ConfigClientSecret,
 			Help: "Box App Client Secret\nLeave blank normally.",
+		}, {
+			Name: "config_json",
```
is `config_json` what users would call it?  Might `jwt_file` or something like that be better?

Also would users ever put it directly in the config file?

Apologies if these are silly questions, I dont know much about JWT!

dbramwell replied:
In the case of Box I think possibly, the "Authenticate with JWT" [tutorial](https://developer.box.com/docs/construct-jwt-claim-manually#section-1-read-json-config) refers to it as `config.json`, and when downloading the file from the box developer page when logged in it saves it as `123456__config.json` (random numbers).

`config.json` does seem rather generic for rclone purposes though. I think `jwt_file` might be misleading as the file doesn't contain the jwt but rather the config to create a jwt assertion which can then be exchanged for an oauth token. Maybe `box_config_file`?

ncw commented on backend/box/box.go:
```diff
@@ -98,6 +128,93 @@ func init() {
+	return key.(*rsa.PrivateKey)
+}
+
+// ConfigJSON defines the shape of a box config.json
```
These type definitions maybe should go in `api/types.go`?

dbramwell replied:
Yes they definitely should, thanks.

ncw commented on lib/jwtutil/jwtutil.go:
```diff
@@ -0,0 +1,90 @@
+	}
+	req, err := http.NewRequest("POST", claims.Aud, nil)
+	if err != nil {
+		log.Fatalf("Failed to create new request: %v", err)
```
This should be returning errors to the user, not exiting out...

So

```go
if err != nil {
    return errors.Wrap(err, "jwtutil: failed to create new request")
}
```

ncw commented on lib/jwtutil/jwtutil.go:
```diff
@@ -0,0 +1,90 @@
+
+	req, err = http.NewRequest("POST", claims.Aud, bytes.NewBuffer([]byte(queryString)))
+	if err != nil {
+		log.Fatalf("Failed to create new request: %v", err)
```
likewise here

ncw commented on lib/jwtutil/jwtutil.go:
```diff
@@ -0,0 +1,90 @@
+	}
+	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
+
+	client := &http.Client{}
```
We should probably pass in a *http.Client here so the caller can pass in [an rclone client](https://github.com/rclone/rclone/blob/master/fs/fshttp/http.go#L188) which will do debuging, rate limiting etc, etc

ncw commented on lib/jwtutil/jwtutil.go:
```diff
@@ -0,0 +1,90 @@
+
+	resp, err := client.Do(req)
+	if err != nil {
+		log.Fatalf("Failed making auth request: %v", err)
```
return error

ncw commented on lib/jwtutil/jwtutil.go:
```diff
@@ -0,0 +1,90 @@
+		log.Fatalf("Failed making auth request: %v", err)
+	}
+	if resp.StatusCode != 200 {
+		log.Fatalf("Failed making auth request: %v", resp.Status)
```
return error

ncw commented on lib/jwtutil/jwtutil.go:
```diff
@@ -0,0 +1,90 @@
+	defer func() {
+		err := resp.Body.Close()
+		if err != nil {
+			fmt.Printf("Error closing response Body: %v", err)
```
return error by naming the return error in the function signature and set it here.

ncw commented on lib/jwtutil/jwtutil.go:
```diff
@@ -0,0 +1,90 @@
+	result := &response{}
+	err = json.NewDecoder(resp.Body).Decode(result)
+	if err != nil || result.AccessToken == "" {
+		log.Fatalf("Failed to get token: %v", err)
```
return error

ncw commented on backend/box/box.go:
```diff
@@ -98,6 +128,93 @@ func init() {
+func getBoxConfig(configFile string) ConfigJSON {
+	jsonData, err := os.Open(configFile)
+	if err != nil {
+		log.Fatalf("Failed to read Box config: %v", err)
```
Ideally this should return the error upwards.

I'm trying to kill off the log.Fatals in the backends so it would be helpful if you could pass errors up and put the log.Fatal in the Config function where it will be easy to fix when you can return an error from there!

ncw commented on backend/box/box.go:
```diff
@@ -98,6 +128,93 @@ func init() {
+	if err != nil {
+		log.Fatalf("Failed to read Box config: %v", err)
+	}
+	byteValue, _ := ioutil.ReadAll(jsonData)
```
error check?

ncw commented on backend/box/box.go:
```diff
@@ -98,6 +128,93 @@ func init() {
 }
 
+func getBoxConfig(configFile string) ConfigJSON {
+	jsonData, err := os.Open(configFile)
```
This open appears to be missing a Close?

You might be better off using ioutil.ReadFile rather than Open/ReadAll/Close

ncw commented on backend/box/box.go:
```diff
@@ -98,6 +128,93 @@ func init() {
+}
+
+func getClaims(boxConfig ConfigJSON, boxSubType string) *jws.ClaimSet {
+	val, _ := jwtutil.RandomHex(20)
```
ignored error?

dbramwell commented:
Great, thanks for the help. I'm very much a go newbie so I've probably completely missed every coding standard! I'll reply to the comments inline.

With regards to adding JWT auth to other backends, from my limited JWT knowledge the Config function in `jwtutil.go` should be generic enough for other backends to use. Each backend that supports JWT should simply have to call it with the correct `ClaimSet` etc. Having said that I'm not sure if any other storage providers support JWT auth in the same way?

dbramwell commented on backend/box/box.go:
```diff
@@ -84,6 +107,13 @@ func init() {
+			Help: "Box App config.json location\nLeave blank normally.",
+		}, {
+			Name:    "box_sub_type",
+			Help:    "'enterprise' or 'user' depending on the type of token being requested.",
```
Thanks, will do

ncw commented:
> Great, thanks for the help. I'm very much a go newbie so I've probably completely missed every coding standard! I'll reply to the comments inline.

I think you've done very well :-)

> With regards to adding JWT auth to other backends, from my limited JWT knowledge the Config function in `jwtutil.go` should be generic enough for other backends to use. Each backend that supports JWT should simply have to call it with the correct `ClaimSet` etc. Having said that I'm not sure if any other storage providers support JWT auth in the same way?

I know drive/gcs support JWT but the SDK has it built in already.

I think that some webdav providers (dput?) can support JWT too.

ncw commented:
Sorry for the radio silence?  How are you getting on with the fixes?  Need help?

dbramwell commented:
Hey. Sorry I was on holiday last week so didn't have any time to work on it. I should get some time tomorrow evening. It all looks quite straight forward, will let you know if I need a hand though, thanks.

ncw commented:
Great - thanks.  Looking forward to merging this :-)

dbramwell commented on backend/box/box.go:
```diff
@@ -73,9 +84,34 @@ func init() {
+				}
+				signingHeaders := getSigningHeaders(boxConfig)
+				queryParams := getQueryParams(boxConfig)
+				client := fshttp.NewClient(fs.Config)
```
Unsure if this is the correct way to create a http client or not

ncw replied:
That's perfect. It will get you an http client with all Rclone's customisations.

dbramwell commented on backend/box/box.go:
```diff
@@ -98,6 +147,74 @@ func init() {
+}
+
 // Options defines the configuration for this backend
 type Options struct {
```
I moved all the types that I created into api/types. I attempted to move other types that already existed, like Options and Fs, but found it quickly escalated into me moving a huge part of the file (lots of Fs function definitions, etc) over. Unsure if this is the correct thing to do

ncw replied:
I usually put just the types that go over the wire in the API in api/types.go.

This file should be mostly struct definitions maybe with a few methods for reading data. Certainly no business logic for rclone .

The point of doing this is so that the definitions can be reused else where.

dbramwell replied:
That makes sense. Does that mean that some of the functions to do with parsing the config file should also go in api/types?

ncw replied:
They could do but I think what you've done is fine :-)

dbramwell commented:
Hey @ncw , sorry for the delay with this.

I think i've addressed all of the issues. I'm unsure if about a couple of points which i've left comments about.

dbramwell commented:
Awesome. Thanks a lot for the guidance.

I've added the snips back in and squashed it all into 3 commits.

ncw commented on docs/content/box.md:
```diff
@@ -31,13 +32,79 @@ Type of storage to configure.
+ 1 / 1Fichier
+   \ "fichier"
+ 2 / A stackable unification remote, which can appear to merge the contents of several remotes
+   \ "union"
```
This still seems to have all the other cloud providers?

This is what it should look like

```
Type of storage to configure.
Choose a number from below, or type in your own value
[snip]
XX / Box
   \ "box"
[snip]
Storage> box
```

dbramwell commented:
Haha, sorry. That makes so much more sense. I think when I first started on this the documentation still had the (out-dated) list of providers in there, hadn't realised that the list had been removed altogether. Still I'm disappointed in myself for not realising what I'd done made absolutely no sense. I blame the fact that it was before 9am!

ncw commented:
> Haha, sorry. That makes so much more sense. I think when I first started on this the documentation still had the (out-dated) list of providers in there, hadn't realised that the list had been removed altogether.

No worries!  I'd forgotten that this PR lived through the changes I made to the docs to remove the other providers.

That looks perfect now, I'll merge it :-)

Thank you very much for doing this.

jacquesvandermerwe commented:
@dbramwell  Thanks for this excellent addition. I have 2 questions, having looked at the code that I cannot figure out.

1. If the box_sub_type is 'user' so that the JWT can do 'As-User'  calls where do we define the user id to use for the 'As-User' call. It isn't in the config setup, is there a switch that is not documented?
2. Does rclone handle expiry for JWT? Since there isn't a refresh_token, what would happen after 60 minutes if the JWT token expires?

Thanks in advance

dbramwell commented:
Hey @jacquesvandermerwe 

1. Good point, reading the Box docs I'm missing that functionality. My organisation doesn't use the As-User stuff. I guess user_id should be an additional config option. We'd then want to add the As-User header in each of the requests if the user_id is set.

2. It doesn't for Box, not sure if it does anywhere else. We get around this by adding the config to our automated process. You can do config in a non-interactive way like this:
`rclone config create box box box_config_file config.json
rclone ls box:/foo`

I guess there's an opportunity here to dig a little deeper into Rclone and make it do the config stuff again when the token expires.

Hope this helps, sorry my answers are basically "It doesn't do what you want it to do"!
